In [2]:
import pandas as pd

dataset = pd.read_csv('./Melbourne_housing.csv')
dataset.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [3]:
dataset.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [4]:
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
               
dataset = dataset[cols_to_use]

In [5]:
dataset.head()

Suburb  Rooms Type Method SellerG             Regionname  \
0  Abbotsford      2    h     SS  Jellis  Northern Metropolitan   
1  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
2  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
3  Abbotsford      3    u     VB  Rounds  Northern Metropolitan   
4  Abbotsford      3    h     SP  Biggin  Northern Metropolitan   

   Propertycount  Distance         CouncilArea  Bedroom2  Bathroom  Car  \
0         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
1         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
2         4019.0       2.5  Yarra City Council       2.0       1.0  0.0   
3         4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
4         4019.0       2.5  Yarra City Council       3.0       2.0  0.0   

   Landsize  BuildingArea      Price  
0     126.0           NaN        NaN  
1     202.0           NaN  1480000.0  
2     156.0          79.0  1035000.0  
3       0.0           NaN        NaN  
4     134.0         150.0  1465000.0

In [6]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [7]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

In [10]:
dataset.isna().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Regionname          3
Propertycount       0
Distance            0
CouncilArea         3
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
Price            7610
dtype: int64

**Drop NA values of Price, since it's our predictive variable we won't impute it**

In [11]:
dataset.dropna(inplace=True)

In [12]:
dataset.shape

(27244, 15)

In [14]:
dataset = pd.get_dummies(dataset, drop_first=True)
dataset.head()

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2         4019.0       2.5       2.0       1.0  1.0     202.0   
2      2         4019.0       2.5       2.0       1.0  0.0     156.0   
4      3         4019.0       2.5       3.0       2.0  0.0     134.0   
5      3         4019.0       2.5       3.0       2.0  1.0      94.0   
6      4         4019.0       2.5       3.0       1.0  2.0     120.0   

   BuildingArea      Price  Suburb_Aberfeldie  ...  \
1      160.2564  1480000.0                  0  ...   
2       79.0000  1035000.0                  0  ...   
4      150.0000  1465000.0                  0  ...   
5      160.2564   850000.0                  0  ...   
6      142.0000  1600000.0                  0  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                    0                                  0   
2                                    0                                  0   
4                                    0                                  0   
5                                    0                                  0   
6                                    0                                  0   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
1                                    0                                      0   
2                                    0                                      0   
4                                    0                                      0   
5                                    0                                      0   
6                                    0                                      0   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                     0                                    0   
2                                     0                                    0   
4                                     0                                    0   
5                                     0                                    0   
6                                     0                                    0   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
1                                    0                                 0   
2                                    0                                 0   
4                                    0                                 0   
5                                    0                                 0   
6                                    0                                 0   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
1                               1                                       0  
2                               1                                       0  
4                               1                                       0  
5                               1                                       0  
6                               1                                       0  

[5 rows x 745 columns]

In [15]:
X = dataset.drop('Price', axis=1)
y = dataset['Price']

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=2)

In [19]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(X_train, y_train)

reg.score(X_train, y_train)

0.6827792395792723

In [20]:
reg.score(X_test, y_test)

0.1385368316164992

**Here training score is 68% but test score is 13.85% which is very low**

Normal Regression is clearly overfitting the data, let's try other models

#### Using Lasso (L1 Regularized) Regression Model

In [23]:
from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg.fit(X_train, y_train)

c:\python_env\datascience\venv_datascience\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1326631542478296.8, tolerance: 796258688308222.6
  model = cd_fast.enet_coordinate_descent(


Lasso(alpha=50, max_iter=100, tol=0.1)

In [25]:
lasso_reg.score(X_test, y_test)

0.6636111369404488

In [26]:
lasso_reg.score(X_train, y_train)

0.6766985624766824

#### Using Ridge (L2 Regularized) Regression Model

In [27]:
from sklearn.linear_model import Ridge
ridge_reg= Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg.fit(X_train, y_train)

Ridge(alpha=50, max_iter=100, tol=0.1)

In [28]:
ridge_reg.score(X_train, y_train)

0.6622376739684328

In [30]:
ridge_reg.score(X_test, y_test)

0.6670848945194959